DXMainClient/Domain/Multiplayer/CnCNet/V3PlayerNegotiator.cs

```diff
@@ -0,0 +1,465 @@
+        finally
+        {
+            onComplete();
+        }
+    }
+
+    // Non-deciders continuously send "Connected" packets to the other player
+    // until they receive a Ping Request. Then they reply with Ping Responses
+    // and await the tunnel choice from the Decider.
+    private async Task PerformNonDeciderNegotiationAsync()
+    {
+        using var cts = CancellationTokenSource.CreateLinkedTokenSource(_negotiationCts.Token);
+        _ = SendConnectedPacketsAsync(cts.Token);
+
+        try
+        {
+            //wait for tuennel choice or negotiation timeout
+            var negotiationTimeout = Task.Delay(NON_DECIDER_TOTAL_TIMEOUT_MS, cts.Token);
+            var completed = await Task.WhenAny(_negotiationCompletionSource.Task, negotiationTimeout);
+
+            if (completed == negotiationTimeout && !_negotiationCompletionSource.Task.IsCompleted)
+            {
+                Logger.Log($"V3TunnelNegotiator: Timeout: No PingRequest received from decider {_remotePlayer.Name} within {NON_DECIDER_TOTAL_TIMEOUT_MS / 1000} seconds.");
+                _negotiationCompletionSource.TrySetResult(false);
+                cts.Cancel();
+                NegotiationComplete?.Invoke(this, EventArgs.Empty);
+            }
+        }
+        catch (OperationCanceledException)
+        {
+            Logger.Log($"V3TunnelNegotiator: Cancelled negotiation with {_remotePlayer.Name}.");
+        }
+    }
+
+    // Send Connected packets every 500ms to tunnels we haven't yet had a ping request from.
+    private async Task SendConnectedPacketsAsync(CancellationToken cancellationToken)
+    {
+        while (!cancellationToken.IsCancellationRequested)
+        {
+            foreach (var tunnel in _tunnels)
+            {
+                var result = _remotePlayer.GetTunnelResult(tunnel);
+                if (result == null || result.ConnectedTimedOut || result.PingRequestReceived)
+                    continue;
+
+                _tunnelHandler.SendPacket(tunnel, _localPlayer.Id, _remotePlayer.Id,
+                    TunnelPacketType.Connected, null);
+
+                if (!result.FirstConnectedSentTime.HasValue)
+                    result.FirstConnectedSentTime = DateTime.UtcNow;
+            }
+
+            await Task.Delay(NON_DECIDER_CONNECTED_INTERVAL_MS, cancellationToken);
+        }
+    }
+
+    //send a ping, wait for response or timeout, next ping...
+    private async Task PerformPingsAsync(CnCNetTunnel tunnel, TunnelTestResult result)
+    {
+        for (int i = 0; i < PINGS_PER_TUNNEL && !_negotiationCts.Token.IsCancellationRequested; i++)
+        {
+            var ping = new PingResult { ID = i, SentTimeTicks = Stopwatch.GetTimestamp() };
+
+            result.PingResults.Add(ping);
+
+            _tunnelHandler.SendPacket(
+                tunnel,
+                _localPlayer.Id,
+                _remotePlayer.Id,
+                TunnelPacketType.PingRequest,
+                BitConverter.GetBytes(i)
```

Comment:
Suggest crash if `BitConverter.IsLittleEndian` does not match our expected value, in case some extremely rare and weird machines break the protocol